evap/staff/tests/test_forms.py

```diff
@@ -554,6 +555,30 @@ def test_staff_can_select_proxy_user(self):
         form = CourseForm(instance=course)
         self.assertIn(proxy_user, form.fields["responsibles"].queryset)
 
+    def test_prevent_contribution_deletion_with_answers(self):
+        """
+        When answers for a contribution already exist, it should not be possible to remove that contribution.
+        """
+        evaluation = baker.make(Evaluation)
+        user1 = baker.make(UserProfile)
+        questionnaire = baker.make(Questionnaire, type=Questionnaire.Type.CONTRIBUTOR)
+        contribution1 = baker.make(
+            Contribution,
+            evaluation=evaluation,
+            contributor=user1,
+            role=Contribution.Role.EDITOR,
+            textanswer_visibility=Contribution.TextAnswerVisibility.GENERAL_TEXTANSWERS,
+            questionnaires=[questionnaire],
+        )
+        baker.make(RatingAnswerCounter, contribution=contribution1)
+
+        contribution_formset = inlineformset_factory(
+            Evaluation, Contribution, formset=ContributionFormSet, form=ContributionForm, extra=1
+        )
+        formset = contribution_formset(instance=evaluation, form_kwargs={"evaluation": evaluation})
+        self.assertFalse(formset.forms[0].instance.can_be_deleted)
+        self.assertTrue(formset.forms[1].instance.can_be_deleted)
```

Comment:
`forms[0]` is `contribution1` and `forms[1]` is the extra instance, right? So we're not really testing that `contribution1` was ever deletable.

In order to make sure that the test really tests the right thing, I usually advocate doing assertions before and after. In this case, that would be:
- Assert that the contribution is deletable
- Add the RatingAnswerCounter
- Assert that the contribution is not deletable anymore.

This is a Forms test, but you're not actually testing any real forms logic. It might make sense to just make it a models test.